src/llama_stack/providers/inline/agents/meta_reference/responses/streaming.py

```diff
@@ -1159,6 +1186,149 @@ async def _process_tools(
+                        try:
+                            final_tools.append(convert_function_tool_choice(chat_tool_names, tool["name"]))
+                        except ValueError as e:
+                            logger.warning(
+                                f"Function tool choice could not be enforced due to the following error: {e}"
+                            )
+                            continue
+                    elif tool["type"] == "mcp":
+                        try:
+                            mcp_tool_choice = convert_mcp_tool_choice(
+                                chat_tool_names, tool["server_label"], self.server_label_to_tools, None
+                            )
+                            if isinstance(mcp_tool_choice, list):
+                                final_tools.extend(mcp_tool_choice)
+                        except ValueError as e:
+                            logger.warning(f"MCP tool choice could not be enforced due to the following error: {e}")
+                            continue
+                    elif tool["type"] == "file_search":
+                        try:
+                            final_tools.append(convert_file_search_tool_choice(chat_tool_names))
+                        except ValueError as e:
+                            logger.warning(
+                                f"File search tool choice could not be enforced due to the following error: {e}"
+                            )
+                            continue
+                    elif tool["type"] == "web_search":
+                        try:
+                            final_tools.append(convert_web_search_tool_choice(chat_tool_names))
+                        except ValueError as e:
+                            logger.warning(
+                                f"Web search tool choice could not be enforced due to the following error: {e}"
+                            )
+                            continue
+                    elif tool["type"] == "custom":
+                        try:
+                            final_tools.append(convert_custom_tool_choice(chat_tool_names, tool["name"]))
+                        except ValueError as e:
+                            logger.warning(f"Custom tool choice could not be enforced due to the following error: {e}")
+                            continue
+                    else:
+                        logger.warning(
+                            f"Unsupported tool type: {tool['type']}, skipping tool choice enforcement for it"
+                        )
+                        continue
+
+                return {
+                    "type": "allowed_tools",
+                    "allowed_tools": {
+                        "tools": final_tools,
+                        "mode": self.ctx.responses_tool_choice.mode,
+                    },
+                }
+
+            elif isinstance(self.ctx.responses_tool_choice, OpenAIResponseInputToolChoiceCustomTool):
```

Comment:
I think we can improve the way this whole function is written perhaps -- these cases feel a bit duplicative? sorry about this being a vague, slightly unactionable comment. but there just feels a bit too much code here for what we want to do.